src/tribol/mesh/MfemData.hpp

```diff
@@ -0,0 +1,1390 @@
+
+#ifndef SRC_MESH_MFEMDATA_HPP_
+#define SRC_MESH_MFEMDATA_HPP_
+
+#include <set>
+#include <vector>
+
+#include "mfem.hpp"
+
+#include "axom/core.hpp"
+#include "redecomp/redecomp.hpp"
+
+#include "tribol/types.hpp"
+#include "tribol/common/Parameters.hpp"
+#include "tribol/mesh/MethodCouplingData.hpp"
+
+namespace tribol
+{
+
+/**
+ * @brief Facilitates transfer to submesh low-order refined mesh
+ *
+ * This class simplifies transferring 1) a variable (e.g. displacement) from a
+ * higher-order grid function on a coarse mesh to a low-order grid function on a
+ * fine mesh and 2) a variable in the dual space (e.g. force) from a low-order
+ * grid function on a fine mesh to a higher-order grid function on a coarse
+ * mesh. The coarse mesh is intended to live on a mfem::ParSubMesh level, but
+ * this is not required to use this class.
+ */
+class SubmeshLORTransfer
+{
+public:
+  /**
+   * @brief Construct a new SubmeshLORTransfer object
+   * 
+   * @param ho_submesh_fes Higher order finite element space on the coarse mesh
+   * @param lor_submesh Refined mesh
+   */
+  SubmeshLORTransfer(
+    mfem::ParFiniteElementSpace& ho_submesh_fes,
+    mfem::ParMesh& lor_submesh
+  );
+
+  /**
+   * @brief Transfers data from a higher-order grid function on a coarse mesh
+   *
+   * Data is transferred to the low-order grid function on a fine mesh that can
+   * be accessed using GetLORGridFn().
+   *
+   * @param ho_src Source higher-order grid function on the coarse mesh
+   */
+  void TransferToLORGridFn(
+    const mfem::ParGridFunction& ho_src
+  );
+
+  /**
+   * @brief Transfers data to a higher-order grid function on a coarse mesh
+   *
+   * Data must be stored in the low-order grid function on a fine mesh accessed
+   * using GetLORGridFn().
+   *
+   * @param ho_dst Destination higher-order grid function on the coarse mesh
+   */
+  void TransferFromLORGridFn(
+    mfem::ParGridFunction& ho_dst
+  ) const;
+
+  /**
+   * @brief Access the local low-order grid function on the refined mesh
+   * 
+   * @return mfem::ParGridFunction& 
+   */
+  mfem::ParGridFunction& GetLORGridFn() { return lor_gridfn_; }
+
+  /**
+   * @brief Access the local low-order grid function on the refined mesh
+   * 
+   * @return const mfem::ParGridFunction& 
+   */
+  const mfem::ParGridFunction& GetLORGridFn() const { return lor_gridfn_; }
+
+private:
+  /**
+  * @brief Create low-order grid function on the refined mesh
+  * 
+  * @param lor_mesh Refined mesh
+  * @param lor_fec Finite element collection to apply to grid function
+  * @param vdim Vector dimension of the grid function
+  * @return mfem::ParGridFunction on lor_mesh, with lor_fec and vdim specified
+  */
+  static mfem::ParGridFunction CreateLORGridFunction(
+    mfem::ParMesh& lor_mesh,
+    std::unique_ptr<mfem::FiniteElementCollection> lor_fec,
+    integer vdim
+  );
+
+  /**
+   * @brief Local low-order grid function on the refined mesh 
+   */
+  mfem::ParGridFunction lor_gridfn_;
```

Comment:
Is this just a generic ParGridFunction in the low-order space that someone can use for whatever they want? If in practice this is LOR coordinates, should we just call it that?

Reply:
This is the grid function on the LOR mesh that the class transfers data to/from.  In the transfer hierarchy, the LOR-level is unique in the sense that nothing outside `MfemData.*pp` will never have to interact with LOR-level data.  Parent level data and submesh level grid functions are used by the host code and redecomp level data is used by Tribol, but LOR level grid functions are just temporary holding places for data in the process of being transferred.  Rather than having another class manage the LOR mesh grid function, I think it makes sense for this class to manage it, since it doesn't need to be interacted with in the host code or Tribol.